### PR TITLE
add gather to ShardedTensor

### PR DESCRIPTION
**Pull request by dracifer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65671

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @gcramer23

Differential Revision: [D31197611](https://our.internmc.facebook.com/intern/diff/D31197611)

### CODE REVIEW

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 1]
```diff
@@ -456,6 +456,35 @@ def test_create_sharded_tensor_with_ones(self):
...
+        h, w = 10, 20
+        sharded_tensor = _sharded_tensor.ones(spec, h, w)
+
+        full_tensor = None
+        if self.rank == 0:
+            full_tensor = torch.zeros(h, w)
+        _sharded_tensor.gather(sharded_tensor, 0, full_tensor)
+
+        if self.rank == 0:
+            self.assertEqual(full_tensor, torch.ones(h, 2))
```
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:484]: Shouldn't this be `torch.ones(h, w)`? Where does the `2` come from?

File: torch/distributed/_sharded_tensor/api.py [hunk 2]
```diff
@@ -335,6 +352,54 @@ def _init_rpc(self):
...
+        if my_rank == dst_rank:
+            for sharded_tensor in gathered_shards:
+                sizes = sharded_tensor.metadata().size
+                if sizes != dst_tensor.size():
+                    raise ValueError(
+                        f"dst_tensor need to have the same size as metadata.size {sizes}"
+                    )
+                dims = len(sizes)
+                pad = [0] * 2 * dims
+                for dim in range(dims):
```
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/api.py:385]: Are we padding each individual shard to be the size of the actual tensor and then just adding it? This is probably wasteful from a memory perspective. Can't we just get a view of `dst_tensor` based on offsets and lengths and just copy the shard into the appropriate position? You can probably use narrow (https://pytorch.org/docs/stable/generated/torch.narrow.html) multiple times to get this?

File: test/distributed/_sharded_tensor/test_sharded_tensor.py (hunk 1 above)
Comment (dracifer) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:484]: typo...

File: torch/distributed/_sharded_tensor/__init__.py [hunk 3]
```diff
@@ -495,3 +495,23 @@ def shard_parameter(
...
+        sharded_tensor (:class `torch.distributed._sharded_tensor.ShardedTensor`): The sharded
+        tensor to gather.
+        dst_rank (int): The rank where full tensor is constructed
+        dst_tensor (:class `torch.Tensor` optional): The output full tensor. Need to and only
+        need to be provided on dst_rank
+
+    Returns:
+        None
+    """
+    return ShardedTensor._gather(sharded_tensor, dst_rank, dst_tensor)
```
[the remark this answers is not in the document]
Comment (dracifer) [reply torch/distributed/_sharded_tensor/__init__.py:517]: to confirm, we can remove this wrapper from __init__? curious what is the rule of thumb about when to put a wrapper in init and when not? thanks!

File: torch/distributed/_sharded_tensor/api.py [hunk 4]
```diff
@@ -335,6 +352,54 @@ def _init_rpc(self):
...
+        dst_tensor: Optional[torch.Tensor] = None,
+    ):
+        my_rank = dist.get_rank(sharded_tensor._process_group)
+        _validate_output_tensor_for_rank(my_rank, dst_rank, dst_tensor)
+
+        shard_tensors = sharded_tensor.local_shards
+
+        gathered_shards = [None for _ in sharded_tensor._process_group.world_size]
+
+        dist.gather_object(
```
[the remark this answers is not in the document]
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:369]: I commented in the internal diff on the same topic. was wonder about the performance overhead here and also wonder about the tradeoff if we flatten locally (larger payload to communicate and more memories). so the local flattening is better?

File: torch/distributed/_sharded_tensor/api.py (hunk 2 above)
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:385]: I discussed this as well in the internal diff. let me know your thoughts

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 5]
```diff
@@ -456,6 +456,35 @@ def test_create_sharded_tensor_with_ones(self):
...
         self.assertEqual(local_shard, torch.ones(expected_h, w))
 
+
+    @with_comms
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_gather_sharded_ones(self):
+        """ Test _sharded_tensor.ones(...) """
+
+        spec = ChunkShardingSpec(
```
[the remark this answers is not in the document]
Comment (dracifer) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:465]: added

File: torch/distributed/_sharded_tensor/api.py [hunk 6]
```diff
@@ -176,6 +177,22 @@ class TensorInitParams(object):
                                  memory_format=torch.contiguous_format,
                                  pin_memory=False))
 
+def _validate_output_tensor_for_rank(my_rank: int, dst_rank: int, dst_tensor: Optional[torch.Tensor]):
+    if dst_rank == my_rank:
+        if not dst_tensor:
+            raise ValueError(
+                "Argument ``dst_tensor`` must be specified on destination rank."
+            )
+        if torch.count_nonzero(dst_tensor):
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_sharded_tensor/api.py:202]: I second @pritamdamania here, we should be able to overwrite any tensor by filling it with the gathered content, there might be cases that user want to use a random tensor and fill it with the gathered content (although this might be rare).

File: torch/distributed/_sharded_tensor/api.py [hunk 7]
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
...
+
+        Args:
+            dst(int): The rank where full tensor is constructed
+            out (:class `torch.Tensor` optional): The output full tensor. Needs to be provided
+            ONLY on `dst`
+        """
+        my_rank = dist.get_rank(self._process_group)
+        full_size = self.metadata().size
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
```
Comment (wanchaol) [inline torch/distributed/_sharded_tensor/api.py:390]: why we need to use `curr_device` here?

File: torch/distributed/_sharded_tensor/api.py [hunk 8]
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
...
+                group=self._process_group,
+            )
+
+        if my_rank == dst:
+            dims = len(full_size)
+            for shards in gathered_shards:
+                assert shards is not None, f"gathered shards cannot be None on {dst}"
+                for shard in shards:
+                    metadata = shard.metadata
+                    tensor = shard.tensor.to(curr_device)
```
Comment (wanchaol) [inline torch/distributed/_sharded_tensor/api.py:411]: there might be case that the tensors gathered together can't fit into a single GPU device right (that's the main reason user might want to shard the tensor). If the whole point of this `gather` call is to gather the content and fill the passed in tensor, should we not care about the device and just copy the content to the `out`?

File: torch/distributed/_sharded_tensor/api.py [hunk 9]
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
...
+
+                    out_narrow_view = out
+                    for dim in range(dims):
+                        out_narrow_view = out_narrow_view.narrow(
+                            dim,
+                            metadata.shard_offsets[dim],
+                            metadata.shard_lengths[dim],
+                        ).contiguous()
+
+                    out_narrow_view.add_(tensor)
```
Comment (wanchaol) [inline torch/distributed/_sharded_tensor/api.py:421]: I guess the main reason you are asserting `zero` for `out` is that you are adding the gathered content to `out`, I think ideally we can just `copy_` it.

**Review by wanchaol — COMMENTED** [review]
@dracifer Thanks for the work! I did an initial pass of review, looks mostly good, left some questions and suggestions in the inlined comments.

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

File: test/distributed/_sharded_tensor/test_sharded_tensor.py [hunk 10]
```diff
@@ -1187,6 +1220,53 @@ def test_create_sharded_tensor_with_ones(self):
         self.assertEqual((5, 5), local_shard.tensor.size())
         self.assertEqual(local_shard.tensor, torch.ones(5, 5))
 
+    @with_comms
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_gather(self) -> None:
+        """ Test _sharded_tensor.gather(...) """
+
+        spec = EnumerableShardingSpec([
```
[the remark this answers is not in the document]
Comment (dracifer) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:1229]: that is the only different. the two tests (all named test_gather) are testing the same function for different sharding strategies so make sure if works for both. they are implemented under different class TestShardedTensorEnumerable and TestShardedTensorChunked, respectively

File: torch/distributed/_sharded_tensor/api.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:202]: got it. I will remove this check then. but just to ensure my understanding correctly, I will not zero the tensor in gather() and add the shards to it directly

File: torch/distributed/_sharded_tensor/api.py [hunk 11]
```diff
@@ -335,6 +367,59 @@ def _init_rpc(self):
...
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
+
+        shard_tensors = self.local_shards()
+        world_size = dist.get_world_size(self._process_group)
+
+        gathered_shards = [None] * world_size
+        # all_gather_object is not efficient. will revise this part once
+        # NCCL support for dist.gather() is ready
+        with torch.cuda.device(curr_device):
```
[the remark this answers is not in the document]
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:398]: I may not have the full context about cpu support. added this as @pritamdamania87 pointed out in prev comment. https://fburl.com/midhqrje. also from test cases, they seems to target on nccl. https://github.com/pytorch/pytorch/blob/master/test/distributed/_sharded_tensor/test_sharded_tensor.py#L269 please let me know if gpu only is expected.

File: torch/distributed/_sharded_tensor/api.py (hunk 9 above)
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:421]: I see. thanks

File: torch/distributed/_sharded_tensor/api.py (hunk 8 above)
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:411]: I think it is a valid concern about size. @wanchaol so we can just do copy_ as you mentioned below. is it correct?

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

File: torch/distributed/_sharded_tensor/api.py [hunk 12]
```diff
@@ -194,7 +222,7 @@ class ShardedTensor(object):
     create_op specified by tensor_init_params.create_op, e.g., torch.ones, or
     torch.empty
 
-    Args:
+    Arg:
```
Comment (bowangbj) [inline torch/distributed/_sharded_tensor/api.py:225]: nit still Args since its multiple?

File: torch/distributed/_sharded_tensor/api.py [hunk 13]
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
...
+            dst(int): The rank where full tensor is constructed. Default is 0.
+            out (:class `torch.Tensor` optional): The output full tensor. Must to be provided
+            ONLY on `dst` rank. Default is None.
+        """
+        my_rank = dist.get_rank(self._process_group)
+        full_size = self.metadata().size
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
+
+        shard_tensors = self.local_shards()
```
Comment (bowangbj) [inline torch/distributed/_sharded_tensor/api.py:391]: nit: local_shards

File: torch/distributed/_sharded_tensor/api.py [hunk 14]
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
...
+
+        TODO: current only supports GPU. Will enable CPU and revise to replace
+              all_gather_object() to gather().
+
+        Args:
+            dst(int): The rank where full tensor is constructed. Default is 0.
+            out (:class `torch.Tensor` optional): The output full tensor. Must to be provided
+            ONLY on `dst` rank. Default is None.
+        """
+        my_rank = dist.get_rank(self._process_group)
```
Comment (bowangbj) [inline torch/distributed/_sharded_tensor/api.py:386]: nit rank

File: torch/distributed/_sharded_tensor/api.py [hunk 15]
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
...
+        full_size = self.metadata().size
+        _validate_output_tensor_for_gather(my_rank, dst, full_size, out)
+        curr_device = torch.device(f"cuda:{my_rank}")
+
+        shard_tensors = self.local_shards()
+        world_size = dist.get_world_size(self._process_group)
+
+        gathered_shards = [None] * world_size
+        # TODO all_gather_object is not efficient. will revise this part once
+        # NCCL support for dist.gather() is ready
```
Comment (bowangbj) [inline torch/distributed/_sharded_tensor/api.py:396]: is below with command necessary?

File: torch/distributed/_sharded_tensor/api.py [hunk 16]
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
...
+        out: Optional[torch.Tensor] = None,
+    ) -> None:
+        """
+        Creates a full :class:`Tensor` on rank `dst` by gathering all sharded tensors.
+
+        The API needs to be called on all ranks in SPMD fashion. All ranks should have
+        the same `dst`. `out` should be a full size zero tensor on `dst` and
+        None on all other ranks.
+
+        TODO: current only supports GPU. Will enable CPU and revise to replace
```
Comment (bowangbj) [inline torch/distributed/_sharded_tensor/api.py:378]: nit: probably add a warning instead?

File: torch/distributed/_sharded_tensor/api.py [hunk 17]
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
...
+            dist.all_gather_object(
+                obj=shard_tensors,
+                object_list=gathered_shards,
+                group=self._process_group,
+            )
+
+        if my_rank == dst:
+            dims = len(full_size)
+            for shards in gathered_shards:
+                assert shards is not None, f"gathered shards cannot be None on {dst}"
```
Comment (bowangbj) [inline torch/distributed/_sharded_tensor/api.py:407]: nit maybe raise exception instead of using assert to make it consistent?

**Review by bowangbj — COMMENTED** [review]
Thanks Bin, looks pretty good, just a few small styling issues

File: torch/distributed/_sharded_tensor/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/__init__.py]: Most of the methods in the init file are around creating a ShardedTensor and not necessarily methods that modify an existing ShardedTensor. If something acts on a instance of ShardedTensor its probably just have it as part of the ShardedTensor class instead of having another wrapper.

File: torch/distributed/_sharded_tensor/api.py [hunk 18]
```diff
@@ -335,6 +363,62 @@ def _init_rpc(self):
...
+    def gather(
+        self,
+        dst: int = 0,
+        out: Optional[torch.Tensor] = None,
+    ) -> None:
+        """
+        Creates a full :class:`Tensor` on rank `dst` by gathering all sharded tensors.
+
+        The API needs to be called on all ranks in SPMD fashion. All ranks should have
+        the same `dst`. `out` should be a full size zero tensor on `dst` and
```
[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/api.py:375]: nit: Out doesn't need to be a zero tensor, rephrase this to "out should be a tensor of the same size as the overall size of the sharded tensor".

File: torch/distributed/_sharded_tensor/api.py (hunk 16 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/api.py:378]: don't put todo in public docs, put this as a comment in the code and in the comment link to the appropriate gh issue.

File: torch/distributed/_sharded_tensor/api.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/api.py:397]: Lets keep the scope of this PR small to address only GPU and then have a follow up to enable CPU as well. Please create a gh issue with the label `sharded_tensor` so we can keep track of this.

File: torch/distributed/_sharded_tensor/api.py (hunk 7 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/api.py:389]: As mentioned above, this won't work for ranks > 8

File: torch/distributed/_sharded_tensor/api.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/api.py:397]: Instead of `curr_device` just use the device for the `out` tensor here.

File: torch/distributed/_sharded_tensor/api.py (hunk 17 above)
Comment (pritamdamania87) [reply torch/distributed/_sharded_tensor/api.py:407]: Yes, this should raise a `RuntimeError`.

File: torch/distributed/_sharded_tensor/api.py [hunk 19]
```diff
@@ -176,6 +177,33 @@ class TensorInitParams(object):
...
+    dst_rank: int,
+    size: torch.Size,
+    dst_tensor: Optional[torch.Tensor],
+) -> None:
+    if dst_rank == my_rank:
+        if dst_tensor is None:
+            raise ValueError(
+                f"Argument ``dst_tensor`` must be specified on destination rank {dst_rank}"
+            )
+        curr_device = torch.device(f"cuda:{my_rank}")
```
[the remark this answers is not in the document]
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:191]: right. I was a bit skeptical for this as well. any API suggestion here?

File: torch/distributed/_sharded_tensor/api.py [hunk 20]
```diff
@@ -176,6 +177,33 @@ class TensorInitParams(object):
...
+        if tuple(size) != (dst_tensor.size()):
+            raise ValueError(
+                f"Argument ``dst_tensor`` have size {tuple(dst_tensor.size())},"
+                f"but should be {tuple(size)}"
+            )
+        if dst_tensor.device != curr_device:
```
[the remark this answers is not in the document]
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:197]: this is mainly checking if dst_tensor is only created on dst_rank as we ask the user to provide dst_tensor only on it.

File: torch/distributed/_sharded_tensor/api.py (hunk 18 above)
[the remark this answers is not in the document]
Comment (dracifer) [reply torch/distributed/_sharded_tensor/api.py:375]: good catch. this is a leftover from previous revision. will update

**Comment by dracifer** [comment]
@dracifer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31197611).

**Review by bowangbj — COMMENTED** [review]
Looks great for me. Please also wait for Pritam/Wanchao's cmt.

**Review by pritamdamania87 — APPROVED** [review]
LGTM, just one additional comment regarding more tests.

**Review by pritamdamania87 — APPROVED** [review]
